### PR TITLE
CU-86dtu8tcn - Add boa test constructor to the documentation

### PR DESCRIPTION
**Pull request by luc10921**

### CODE REVIEW

**Comment by melanke** [comment]
Task linked: [CU-86dtu8tcn Add boa test constructor to the documentation](https://app.clickup.com/t/86dtu8tcn)

File: docs/source/testing-and-debugging.md [hunk 1]
```diff
@@ -43,71 +43,79 @@ from boa3.boa3 import Boa3
...
 
-## Neo Test Runner
+## boa test constructor
 
 ### Downloading
 
-Install [Neo-Express](https://github.com/neo-project/neo-express#installation) and [Neo Test Runner](https://github.com/ngdenterprise/neo-test#neo-test-runner).
+Install [boa-test-constructor](https://pypi.org/project/boa-test-constructor/) with pip. We use this extension to run an isolated
+test environment for smart contracts with a neo-go node. When installing boa-test-constructor, [neo-mamba](https://dojo.coz.io/neo3/mamba/index.html)
+will be installed too.
```
Comment (meevee98) [inline docs/source/testing-and-debugging.md:52]: It is worth mentioning that the packages can be installed directly when neo3-boa is installed by using `pip install neo3-boa[test]`

File: docs/source/testing-and-debugging.md [hunk 2]
```diff
@@ -43,71 +43,79 @@ from boa3.boa3 import Boa3
...
+the test environment, you'll need to override the `setUpClass` method from `SmartContractTestCase`. This method is 
+synchronous, so if you need to set up asynchronous tasks, you can create another async method and use it int the 
+`asyncio.run` method from `asyncio`. Common operations would be: creating accounts, deploying the smart contract, 
+selecting your "main" smart contract, and transferring GAS to the new accounts.
 
-Create a Python Script, import the NeoTestRunner class, and define a function to test your smart contract. In this
-function you'll need a NeoTestRunner object, which takes the path of your Neo-Express network configuration file as an
-argument to set up the test environment.
+Then, create functions to test the expected behavior of your smart contract. To invoke or test invoke your smart 
+contract, use the `call` method from `SmartContractTestCase`. The two positional parameters are the name of the method
```
Comment (meevee98) [inline docs/source/testing-and-debugging.md:63]: the user reading this documentation may not know the difference between invoke and test invoke, you can just write "To invoke your smart contract [...]" here

File: docs/source/testing-and-debugging.md [hunk 3]
```diff
@@ -43,71 +43,79 @@ from boa3.boa3 import Boa3
...
+synchronous, so if you need to set up asynchronous tasks, you can create another async method and use it int the 
+`asyncio.run` method from `asyncio`. Common operations would be: creating accounts, deploying the smart contract, 
+selecting your "main" smart contract, and transferring GAS to the new accounts.
 
-Create a Python Script, import the NeoTestRunner class, and define a function to test your smart contract. In this
-function you'll need a NeoTestRunner object, which takes the path of your Neo-Express network configuration file as an
-argument to set up the test environment.
+Then, create functions to test the expected behavior of your smart contract. To invoke or test invoke your smart 
+contract, use the `call` method from `SmartContractTestCase`. The two positional parameters are the name of the method 
+you want to invoke and a list of its arguments. The keyword parameters are the return type, a list of signing accounts,
```
Comment (meevee98) [inline docs/source/testing-and-debugging.md:64]: ensure that methodName, args and returnType are mandatory

File: docs/source/testing-and-debugging.md [hunk 4]
```diff
@@ -43,71 +43,79 @@ from boa3.boa3 import Boa3
...
-the compiled smart contract, the smart contract's method, and the arguments if necessary. 
-This call doesn't return the result directly, but includes it in a queue of invocations. To execute all the invocations
-set up, call the method `execute()`. Then assert the result of your invoke to see if it's correct.
-
-Note that `invoke.result` won't be set if the execution fails, so you should also assert if `runner.vm_state` is valid 
-for your test case.
+To persist an invocation, use the `signing_accounts` parameter to pass a list of signing accounts when calling the 
+smart contract. If you don't pass it, then it will always be a test invoke. The `signers` parameter can be used 
+alongside the `signing_accounts` if you want to change the witness scope of the invocation, or by itself if you want to
+test invoke but also define the signers of the transaction.
```
Comment (meevee98) [inline docs/source/testing-and-debugging.md:71]: we are not going to explain what signers and witnesses scopes are in this file, can you link the neo documentation related to signers here to guide the user if it wants more information?

File: docs/source/testing-and-debugging.md [hunk 5]
```diff
@@ -43,71 +43,79 @@ from boa3.boa3 import Boa3
...
+
+        asyncio.run(cls.asyncSetupClass())
+
+    @classmethod
+    async def asyncSetupClass(cls) -> None:
+        # this `transfer` method already uses the correct amount of decimals for the token
+        await cls.transfer(GAS, cls.genesis.script_hash, cls.user1.script_hash, 100)
+        
+        # the smart contract I'm deploying is hello_world_with_deploy.py from the "Neo Methods" https://dojo.coz.io/neo3/boa/getting-started.html#neo-methods 
+        cls.contract_hash = await cls.deploy("./smart_contract.nef", cls.genesis)
```
Comment (meevee98) [inline docs/source/testing-and-debugging.md:106]: I think these code snippets should be along the texts that explain what to do, the text by itself does not help much without code to look into
If the user if following the documentation I don't think its test is going to look like this as the text before it suggests, there is mention to initiating accounts, but is not explicit that in this tutorial we're setting a user test account.

Comment (meevee98) [inline docs/source/testing-and-debugging.md:106]: It's worth mentioning that the test case run on a local chain that is going to reset each time the tests are executed, that's why the contract needs to be deployed and the user balance needs to me initialized

File: docs/source/testing-and-debugging.md [hunk 6]
```diff
@@ -43,71 +43,79 @@ from boa3.boa3 import Boa3
...
+    user1: Account
+    
+    # if this variable is set, then this contract hash will be called whenever you don't use the `target_contract` parameter on the `call` method
+    contract_hash: types.UInt160
+
+    @classmethod
+    def setUpClass(cls) -> None:
+        super().setUpClass()
+        # you can name the account whatever you want, but the password needs to be "123"
+        cls.user1 = cls.node.wallet.account_new("123", "alice")
```
Comment (meevee98) [inline docs/source/testing-and-debugging.md:95]: just for clarity sake, i think it's better to put the label first
```suggestion
        cls.user1 = cls.node.wallet.account_new(label="alice", password="123")
```

File: docs/source/testing-and-debugging.md [hunk 7]
```diff
@@ -43,71 +43,79 @@ from boa3.boa3 import Boa3
...
+    genesis: Account
+    user1: Account
+    
+    # if this variable is set, then this contract hash will be called whenever you don't use the `target_contract` parameter on the `call` method
+    contract_hash: types.UInt160
+
+    @classmethod
+    def setUpClass(cls) -> None:
+        super().setUpClass()
+        # you can name the account whatever you want, but the password needs to be "123"
```
Comment (meevee98) [inline docs/source/testing-and-debugging.md:94]: is there a reason that the password needs to be "123"? is it a boa-test-case issue?

File: docs/source/testing-and-debugging.md [hunk 8]
```diff
@@ -43,71 +43,79 @@ from boa3.boa3 import Boa3
...
 
-Before writing your tests, make sure you have a Neo-Express network for local tests.
-If you do not yet have a local network, open a terminal and run `neoxp create`.
-Please refer to [Neo-Express documentation](https://github.com/neo-project/neo-express/blob/master/docs/command-reference.md#neoxp-create)
-for more details of how to configure your local network. 
+Create a Python Script, import the `SmartContractTestCase` class, and create a test class that inherits it. To set up
+the test environment, you'll need to override the `setUpClass` method from `SmartContractTestCase`. This method is 
+synchronous, so if you need to set up asynchronous tasks, you can create another async method and use it int the 
+`asyncio.run` method from `asyncio`. Common operations would be: creating accounts, deploying the smart contract, 
+selecting your "main" smart contract, and transferring GAS to the new accounts.
```
Comment (meevee98) [inline docs/source/testing-and-debugging.md:60]: it is worth mentioning why the tests must be async:
the tests use a local chain to run the invocations and the communication with the chain is asynchronous, so anything that needs to interact with it (deploy, token transfers, vote, etc) must be awaited

**Review by meevee98 — CHANGES_REQUESTED** [review]

**Review by meevee98 — APPROVED** [review]

**Pull request merged**